crates/bgp-pkt/src/bgp_ls.rs

```diff
@@ -0,0 +1,1273 @@
+    /// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    /// |              Type             |             Length            |
+    /// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    /// | Prefix Length | IP Prefix (variable)                         //
+    /// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    /// ```
+    /// see [RFC7752 Section 3.3.2.3](https://www.rfc-editor.org/rfc/rfc7752#section-3.2.3.2)
+    IpReachabilityInformation(IpReachabilityInformationData),
+}
+
+impl BgpLsPrefixDescriptorTlv {
+    pub fn get_type(&self) -> iana::BgpLsPrefixDescriptorTlvType {
+        match self {
+            BgpLsPrefixDescriptorTlv::MultiTopologyIdentifier(..) => {
+                iana::BgpLsPrefixDescriptorTlvType::MultiTopologyIdentifier
+            }
+            BgpLsPrefixDescriptorTlv::OspfRouteType(_) => {
+                iana::BgpLsPrefixDescriptorTlvType::OspfRouteType
+            }
+            BgpLsPrefixDescriptorTlv::IpReachabilityInformation(_) => {
+                iana::BgpLsPrefixDescriptorTlvType::IpReachabilityInformation
+            }
+        }
+    }
+}
+
+/// ```text
+///  0                   1                   2                   3
+///  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+/// +-+-+-+-+-+-+-+-+
+/// |  Protocol-ID  |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |                           Identifier                          |
+/// |                            (64 bits)                          |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// //                Local Node Descriptors (variable)            //
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// ```
+/// see [RFC7752 Section 3.2](https://www.rfc-editor.org/rfc/rfc7752#section-3.2)
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(feature = "fuzz", derive(arbitrary::Arbitrary))]
+pub struct BgpLsNlriNode {
+    pub protocol_id: BgpLsProtocolId,
+    pub identifier: u64,
+    pub local_node_descriptors: BgpLsNodeDescriptorTlv,
+}
+
+/// ```text
+///  0                   1                   2                   3
+///  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |              Type             |             Length            |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |                                                               |
+/// //              Node Descriptor Sub-TLVs (variable)            //
+/// |                                                               |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// ```
+#[derive(Display, Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(feature = "fuzz", derive(arbitrary::Arbitrary))]
+pub enum BgpLsNodeDescriptorTlv {
```

Comment:
I would drop Tlv from all names, it's redundant

Reply:
yeah naming things is hard🙂 currently i used the names from the rfc because it helps understanding what the structs refer to. any suggestion for what to call the tlvs/sub-tlvs without being confusing or including "tlv"?

Reply:
I've to read more of the RFC to be able to suggest something more meaningful. I had similar issue with the extended communities, see https://github.com/NetGauze/NetGauze/blob/main/crates/bgp-pkt/src/community.rs